### ISSUE TITLE
Subscription/Registration meta-api

### ISSUE THREAD

**Issue by darkl** [tavendo/WAMP#176]
This is related to the current implementation of crossbar (see [here](http://crossbar.io/docs/Subscription-Meta-Events-and-Procedures/)), which I guess will be close to the spec.

Currently one can subscribe to wamp.subscription.on_subscribe/wamp.subscription.on_unsubscribe and get all notifications about added/removed clients to all subscriptions of the router.

I suggest to have instead the subscription id specified in the topic, i.e. wamp.subscription.on_subscribe.1099266751 and wamp.subscription.on_unsubscribe.1099266751, and something similar for registrations.

This allows to subscribe only to events of a specific subscription. Of course you can still subscribe to all subscriptions using match: 'prefix'.

**Comment by ecorm** [comment]
@darkl , a similar feature had already been proposed in [Subscriber Meta Events](https://github.com/tavendo/WAMP/blob/master/spec/advanced/subscription-meta-api.md#subscriber-meta-events). It proposed to use an "metatopics" option in SUBSCRIBE messages.

In my PR #179 , I had proposed two meta-events that would replace the `SUBSCRIBE.Options.metatopics` idea. Namely, [`wamp.subscription.on_subscribe_to.<topic>`](https://github.com/ecorm/WAMP/blob/ecorm-issue-157/spec/advanced.md#wampsubscriptionon_subscribe_totopic) and [`wamp.subscription.on_unsubscribe_from.<topic>`](https://github.com/ecorm/WAMP/blob/ecorm-issue-157/spec/advanced.md#wampsubscriptionon_unsubscribe_fromtopic).

I had chosen `wamp.subscription.on_subscribe_to.<topic>` so that it can't possibly get confused with the more general `wamp.subscription.on_subscribe` (especially if you start thinking about pattern matching).

Would `wamp.subscription.on_subscribe_to.<subscription_id>` make more sense than `wamp.subscription.on_subscribe_to.<topic>`? The latter can result in some pretty long, convoluted URIs.

**Comment by darkl** [comment]
I think subscription ids are better for this purpose, since the scenario I
imagine is users subscribe
to wamp.subscription.on_create/wamp.subscription.on_delete and then
subscribe
to wamp.subscription.on_subscribe_to.<subscription_id>/wamp.subscription.on_unsubscribe_to.<subscription_id>
based on the created subscription.

**Comment by ecorm** [comment]
If one wanted behavior similar to `wamp.subscription.on_subscribe_to.<topic>`, she could use `wamp.subscription.lookup` (which returns a subscription id), and then subscribe to `wamp.subscription.on_subscribe_to.<subscription_id>`.

However, if nobody subscribed to `<topic>` yet, I don't see how the above workaround would work.

**Comment by darkl** [comment]
Yeah, but we need to notice that topic uri isn't 1-1 with subscription ids.
(One can subscribe to a topic with match="prefix" and match="wildcard" and
these will yield different subscription ids)

So I'm not sure if it makes sense to subscribe about updates of a topic
uri, maybe to updates about a topic uri and subscribeOptions, but I guess
it's simpler to just subscribe by subscription id.

**Comment by oberstet** [comment]
yeah, as @darkl mentions, `wamp.subscription.on_subscribe_to.<topic>` doesn't work, because while a publication is always to a concrete topic, a subscription can also be for a pattern.

`wamp.subscription.on_subscribe_to.<subscription_id>` doesn't have that problem.

The topic should probably be named differently though: `wamp.subscription.<subscription_id>.on_subscribe`.

The reason is to stay in the URI scheme "object"."event". In the `wamp.subscription.on_subscribe` topic, the object is the category `wamp.subscription` (any subscription), whereas in `wamp.subscription.<subscription_id>.on_subscribe`, the object is `wamp.subscription.<subscription_id>` - a single concrete subscription.

@ecorm 

     CALL: wamp.subscription.lookup -> subscription_id
     SUB: wamp.subscription.<subscription_id>.on_subscribe

looks fine to me.

It may seem weird to some that `lookup` would return a subscription ID even for a topic/pattern/subflags that isn't subscribed to by anyone currently. Note that we are talking about CB!

We have to be very careful not introduce CB implementation artefacts into the spec.

E.g. with CB, a subscribe to the same topic/pattern will always return the same subscription ID. For 1 client doing 2 subsequent subscribes or for 2 clients.

BUT: this is an implementation details. The WAMP BP doesn't specify the exact nature of subscription IDs. In fact, a Broker is allowed to hand out different subscription IDs to all clients, even when they are subscribed to the same topic/pattern. The only thing it must do: send events with the previously handed out subscription IDs when events are published to the originally subscribed topic/pattern.

> However, if nobody subscribed to <topic> yet, I don't see how the above workaround would work.

Yep, that's an implementation problem. In CB, I'd just allow subscriptions with 0 subscribers but >0 meta-subscribers to exist "ghost like" for these kind of situations.

**Comment by darkl** [comment]
Maybe we should introduce a meta-topic that tells us when is a subscription
corresponding to a given topic is created. Something like "
wamp.subscription.on_create.com.myapp.topic1" so users can be aware when
they can subscribe to wamp.subscription.<subscription_id>.on_subscribe.

Of course this has the famous race-condition issue that until you subscribe
to the inner meta-topic, events will already be published (And therefore
you'll need to call an rpc).

**Comment by ecorm** [comment]
> We have to be very careful not introduce CB implementation artefacts into the spec.

You should be on the lookout for that when reviewing my meta API PR. You might have to go in and edit a few implementation details that leaked into the spec.

> The reason is to stay in the URI scheme "object"."event". In the `wamp.subscription.on_subscribe topic`, the object is the category `wamp.subscription` (any subscription), whereas in `wamp.subscription.<subscription_id>.on_subscribe`, the object is `wamp.subscription.<subscription_id>` - a single concrete subscription.

That makes perfect sense. Thanks for the explanation.

> Maybe we should introduce a meta-topic that tells us when is a subscription corresponding to a given topic is created. Something like `"wamp.subscription.on_create.com.myapp.topic1"` so users can be aware when they can subscribe to `wamp.subscription.<subscription_id>.on_subscribe`.

Would it be too much to ask a client to subscribe to the generic `wamp.subscription.on_create` meta-event and simply check the `"uri"` property of the emitted event argument?

**Comment by darkl** [comment]
I think it's fine to let the user check himself the "uri" property, but this might spam some users too much (but in my opinion this should be fine).

**Comment by oberstet** [comment]
I changed the issue title to better reflect what we have been discussing above.

The use case is:

* a user wants to receive meta API events for an individual subscription/registration
* the current API only allows to receive meta API events for _all_ subscriptions/registrations
* and the user does want to **save resources** processing stuff the user isn't interested in

@darkl is that what you had in mind?

**Comment by darkl** [comment]
Yeah, exactly.

**Comment by konsultaner** [comment]
I would like to implement this feature into my implementation. I was thinking about all the comments done in this issue.

@ecorm  wrote
> However, if nobody subscribed to <topic> yet, I don't see how the above workaround would work.

@darkl  wrote
> Yeah, but we need to notice that topic uri isn't 1-1 with subscription ids.
(One can subscribe to a topic with match="prefix" and match="wildcard" and
these will yield different subscription ids)

So this is obviously a point that needs to be solved.

if a topic uri is used for both a prefix and an simple match, a router is not able to decide what to do. But without an initial subscription no subscription id is available.

Picking up @oberstet notation, this notation would address both issues:

`wamp.subscription.<topic>.plain.on_subscribe`
`wamp.subscription.<topic>.prefix.on_subscribe`
`wamp.subscription.<topic>.wildcard.on_subscribe`

or 

`wamp.subscription.<subscription_id>.on_subscribe`

to address a concrete subscription id

This would also keep implementers from having trouble with race conditions as mentioned by @darkl 
> Of course this has the famous race-condition issue that until you subscribe
to the inner meta-topic, events will already be published (And therefore
you'll need to call an rpc).

**Comment by oberstet** [comment]
> wamp.subscription.<topic>.plain.on_subscribe

fwiw, as explained above, I am -1 on putting `<topic>` (the application level topic)` into the WAMP meta topic URI.

it also wouldn't work practically: wildcard URI patterns allow empty URI components (URI parts), which leads to ambiguity when interpreting the "outer" WAMP meta topic URI. 

IMO, the only data piece that makes sense to be part of the meta topic URI is the `subscription_id`, and that also cohesively fits the existing design is

```
wamp.subscription.<subscription_id>.on_subscribe
```

**Comment by konsultaner** [comment]
> it also wouldn't work practically: wildcard URI patterns allow empty URI components (URI parts), which leads to ambiguity when interpreting the "outer" WAMP meta topic URI.

@oberstet good point. this is an edge-case that would need the existance of `wamp.subscription.<subscription_id>.on_subscribe`

What about `wamp.subscription.lookup` **MUST** (for this feature)  return an subscription id. The router would have to create an empty subscription as CB handles it? Sounds good to me to be honest. I would create the subscription if is has been looked or subscribed to. I would delete the subscription If it has never been looked up an no one has subscribed to it. If it has been looked up I would keep it forever.

>E.g. with CB, a subscribe to the same topic/pattern will always return the same subscription ID. For 1 client doing 2 subsequent subscribes or for 2 clients.

@oberstet Isn't that part of the specs anyway? I implemented it this way.

**Comment by oberstet** [comment]
> What about wamp.subscription.lookup MUST (for this feature) return an subscription id.

The spec is explicit about this https://wamp-proto.org/_static/gen/wamp_latest_ietf.html#wampsubscriptionlookup

When there is no one subscribed, `null` is returned.

IOW: `wamp.subscription.lookup` is _idempotent_ wrt to the broker state (eg the set of actively managed subscriptions).

CB implements exactly this behavior.

Rgd "always returning the same ID": CB does indeed currently return the same ID, but that is an implementation detail, pls see my comment https://github.com/wamp-proto/wamp-proto/issues/176#issuecomment-136505015

**Comment by konsultaner** [comment]
@oberstet To be honest, I don't see this (not yet) feature work smoothly without extending the `Subscribe.Options`. Are you totaly against extending the options?

**Comment by konsultaner** [comment]
I started implementing this (not yet) feature.

>IMO, the only data piece that makes sense to be part of the meta topic URI is the subscription_id, and >that also cohesively fits the existing design is
>
>    wamp.subscription.<subscription_id>.on_subscribe

@oberstet This obviously only works if one would implement the subscriptionId handlich as CB does. Thats why I came up with this and it seems to work for all issues mantioned above by @ecorm and @darkl 

- I extended the Subscribe.Options with the field `meta_topic`. 
- If a subscriber subscribes to `wamp.subscription.on_subscribe` and `meta_topic` is empty ( == null in my case) the regular meta api behavior is used. 
- If `meta_topic` is a valid uri, the dealer creates a new unique subscriptionId to identify the `meta_topic` inside the dealer implementation
- The `Subscribe.Options.match` is used to for `meta_topic` and not for `Subscribe.topic` anymore (this is how I needed it, maybe another handling is better)
- If a subscriber subscribes a watched `meta_topic` (or a wildcard or prefix match) the dealer will find the meta subscribers subscriptionId and publishes the event to the meta subscriber

it also works with the other `meta.subscription.X`

**Comment by konsultaner** [comment]
@oberstet 
>E.g. with CB, a subscribe to the same topic/pattern will always return the same subscription ID. For 1 >client doing 2 subsequent subscribes or for 2 clients.
>
>BUT: this is an implementation details. The WAMP BP doesn't specify the exact nature of subscription >IDs. In fact, a Broker is allowed to hand out different subscription IDs to all clients, even when they are >subscribed to the same topic/pattern.

If this is true, does `wamp.registration.lookup` and `wamp.subscription.lookup` return the first found `subscriptionId` or a list?

**Comment by oberstet** [comment]
@konsultaner 

* you should probably prefix `meta_topic` and `match` with `x_` as it is a non-standard attribute (not in WAMP) - also, this is OT to the this issue here
* `lookup` takes an `options` kwarg which can be used to specify the matching policy to be looked up
* I can see how the former is still insufficient to resolve to a unique subscription/registration ID should the router implementation not behave like CB, but might use many/different IDs (even for the same matching policy), and hence there is indeed sth missing in the spec: either we should say a) "first ID found" or b) return a list of IDs (possibly), or c) follow what CB does ..

**Comment by konsultaner** [comment]
@oberstet 
> you should probably prefix meta_topic and match with x_ as it is a non-standard attribute (not in WAMP) - also, this is OT to the this issue here

Good point, i will prefix it. 

I don't think this is off topic. Isn't it a good input to others, how I solved the implementation without having to deal with any issues described in this thread?

**Comment by oberstet** [comment]
> how I solved the implementation without having to deal with any issues described in this thread?

There are no issues with what I proposed.
